azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_mqtt.py

```diff
@@ -176,3 +210,32 @@ def _on_message_received(self, topic, payload):
             stage=self,
             event=pipeline_events_mqtt.IncomingMQTTMessageEvent(topic=topic, payload=payload),
         )
+
+    def _on_unexpected_connection(self):
+        """
+        Handler that gets called by the transport when it connections.  This handler only
+        gets called when it connects outside of the normal connect/reconnect logic.
+        """
+        logger.warn("an unexpected connection occured")
+        self.on_connected()
+
+    def _on_unexpected_connection_failure(self, cause):
+        """
+        Handler that gets called by the transport when a connection fails.  This handler only
+        gets called when a connection fails outside of the normal connect/reconnect logic,
+        which should not happen, so the only real route of action is to pass the failure up
+        to the unhandled error handler
+        """
+        logger.error("an unexpected connection failure occured: {}".format(cause))
+        self.pipeline_root.unhandled_error_handler(cause)
+
+    def _on_unexpected_disconnection(self, cause):
```

Comment:
>_on_unexpected_disconnection [](start = 8, length = 28)

Another issue here.  An unexpected disconnection is easy, we use "on_disconnected" to pass the event up, but do we do anything special with the error that indicates why we disconnected?  I don't think we need to handle unexpected disconnections differently than expected disconnections.  Or do we?  Also, my code passes the cause up as a "unhandled error", but this could cause confusion.  Maybe I pass up an "unexpected disconnection" error wrapping the cause of the unexpected disconnection?

Reply:
can we call `on_disconnected()` ?

---
In reply to: [302278864](https://github.com/Azure/azure-iot-sdk-python-preview/pull/136#discussion_r302278864) [](ancestors = 302278864)

Reply:
I agree - our error story should be as clear as possible. Unexpected disconnection would be preferable.

---
In reply to: [302278864](https://github.com/Azure/azure-iot-sdk-python-preview/pull/136#discussion_r302278864) [](ancestors = 302278864)